src/messages/permissions/Authentication.ts

```diff
@@ -1,85 +1,133 @@
 import express, { Express } from 'express';
 import DiscordClient from '../../DiscordClient';
 import { URLSearchParams } from 'url';
-import { requestWrapper as request } from '../../common';
+import Common, { requestWrapper as request } from '../../common';
 import { RequestOptions } from 'https';
 import { Server } from 'http';
-import fs from 'fs';
-import Common from '../../common';
 import Settings, { SettingsJSON } from '../../Settings';
+import { GuildInfo } from '../../settings/SettingsDB';
+import GuildSettings from '../../settings/GuildSettings';
+import { Guild } from 'discord.js';
 
-interface TokenResponse {
+export interface TokenResponse {
   access_token: string;
   expires_in: number;
   scope: string;
   token_type: string;
   refresh_token: string;
 }
 
+export type MaybeTokenResponse = TokenResponse | { error: string } | void;
+
 /**
  * Creates an express app for the user to authorize the bot to allow changing command permissions
  */
-export default class Authentication {
-  #app: Express = express();
-  #server: Server;
+export default abstract class Authentication {
```

Comment:
This class has a lot of functions. Sort them and, if possible, extract into subclasses